crates/partition-store/src/snapshots.rs

```diff
@@ -0,0 +1,78 @@
+use restate_types::identifiers::{PartitionId, PartitionKey};
+use restate_types::logs::Lsn;
+use rocksdb::LiveFile;
+use serde::{Deserialize, Serialize};
+use serde_with::{serde_as, DeserializeAs, SerializeAs};
+use std::ops::RangeInclusive;
+use std::path::PathBuf;
+
+/// Partition snapshot description.
+#[serde_as]
+#[derive(Debug, Serialize, Deserialize)]
+pub struct PartitionSnapshotMetadata {
+    /// Restate cluster name which produced the snapshot.
+    pub cluster_name: String,
+
+    // todo(pavel): the partition table version may be useful for sanity check on restore
+    /// Restate partition id.
+    pub partition_id: PartitionId,
+
+    /// The partition key range that the partition processor which generated this snapshot was
+    /// responsible for, at the time the snapshot was generated.
+    pub key_range: RangeInclusive<PartitionKey>,
+
+    /// The minimum LSN guaranteed to be applied in this snapshot. The actual
+    /// LSN may be >= [minimum_lsn].
+    pub min_applied_lsn: Lsn,
+
+    /// The RocksDB SST files comprising the snapshot.
+    #[serde_as(as = "Vec<SnapshotSstFile>")]
+    pub files: Vec<LiveFile>,
+}
+
+/// A locally-stored partition snapshot.
+#[derive(Debug)]
+pub struct LocalPartitionSnapshot {
+    pub base_dir: PathBuf,
+    pub min_applied_lsn: Lsn,
+    pub db_comparator_name: String,
+    pub files: Vec<LiveFile>,
+}
+
+/// RocksDB SST file that is part of a snapshot. Serialization wrapper around [LiveFile].
+#[serde_as]
+#[derive(Serialize, Deserialize)]
+#[serde(remote = "LiveFile")]
+pub struct SnapshotSstFile {
+    pub column_family_name: String,
+    pub name: String,
+    pub directory: String,
+    pub size: usize,
+    pub level: i32,
+    #[serde_as(as = "Option<serde_with::hex::Hex>")]
```

Comment:
Curious, why Hex and not Bytes?

Reply:
You'll laugh, I was anchored on hex from staring at the C++ LiveFile struct in the debugger view projecting that it might be nicer for humans to look at. No real preference - `Option<Bytes>` is just as good for me.

Reply:
Actually, I do have a strong preference for it not being an array, i.e. just the default `Bytes` representation. Since this metadata is likely to be seen by a human in its raw form, it's nice to be pretty-printed and I think there's a big difference between a 15-20 line array like this:

```
17        "start_key": [
  1         100,
  2         101,
  3         0,
  4         0,
  5         0,
  6         0,
  7         0,
  8         0,
  9         0,
 10         0,
 11         1,
 12         4,
 13         83,
 14         69,
 15         76,
 16         70
 17       ],
```

and a shorter hex or base64 representation. Hex made more sense as slightly easier to interpret - if I saw this and was trying to make sense of the key without a Restate-specific tool to render the key, my first reaction would be to hex dump it:

```
"start_key": "ZGUAAAAAAAAAAQEEU0VMRg==",
```

So that was my thought process - what would be your first choice? :-)

Reply:
I made this base64 in the end - but still interested to know what others think!

Reply:
The positive point for hex is that we can compare the values to the output from rocks' `ldb` CLI tool

Reply:
Ok, that sells me on hex!